src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/seed/cas1/ApprovedPremisesRoomsSeedFromXLSXJob.kt

```diff
@@ -0,0 +1,210 @@
+  private val roomRepository: RoomRepository,
+  private val bedRepository: BedRepository,
+  private val siteSurvey: SiteSurvey,
+) : ExcelSeedJob() {
+  private val log = LoggerFactory.getLogger(this::class.java)
+  override fun processDataFrame(dataFrame: DataFrame<*>, premisesId: UUID) {
+    findExistingPremisesOrThrow(premisesId)
+
+    // build valid lists for rooms, characteristics and beds
+    var rooms = buildRooms(dataFrame, premisesId)
+    val characteristics = buildCharacteristics(dataFrame)
+
+    // add the characteristics to the rooms if they exist
+    rooms.forEach {
+      characteristics[it.code]?.let { roomCharacteristics -> it.characteristics.addAll(roomCharacteristics.toList()) }
+    }
+
+    // update the rooms
+    rooms = createOrUpdateRooms(rooms)
+
+    val beds = buildBeds(dataFrame, rooms)
+    createBedsIfNotExist(beds)
+  }
+
+  private fun buildRooms(dataFrame: DataFrame<*>, premisesId: UUID): MutableList<RoomEntity> {
+    val rooms = mutableListOf<RoomEntity>()
+
+    for (i in 1..<dataFrame.columnsCount()) {
+      val roomAnswers = dataFrame.getColumn(i)
+      val room = createRoom(premisesId, roomAnswers.name, roomAnswers[0].toString())
+
+      rooms.add(room)
+    }
+    return rooms
+  }
+
+  private fun buildCharacteristics(dataFrame: DataFrame<*>): MutableMap<String, MutableList<CharacteristicEntity>> {
+    var premisesCharacteristics = mutableMapOf<String, MutableList<CharacteristicEntity>>()
+
+    siteSurvey.questionToCharacterEntityMapping.forEach { (question, characteristic) ->
+      val rowId = dataFrame.getColumn(0).values().indexOf(question)
+
+      if (rowId == -1) throw SiteSurveyImportException("Characteristic question '$question' not found on sheet Sheet3.")
+
+      for (colId in 1..<dataFrame.columnsCount()) {
+        val roomCode = dataFrame.getColumn(colId).name
+        val answer = dataFrame[rowId][colId].toString().trim()
+
+        if (answer.equals("yes", true)) {
+          premisesCharacteristics.computeIfAbsent(roomCode) { mutableListOf() }.add(characteristic!!)
+        } else if (!answer.equals("no", true) && !answer.equals("N/A", true)) {
+          throw SiteSurveyImportException("Expecting 'yes' or 'no' for question '$question' but is '$answer' on sheet Sheet3.")
+        }
+      }
+    }
+    return premisesCharacteristics
+  }
+
+  private fun buildBeds(dataFrame: DataFrame<*>, rooms: MutableList<RoomEntity>): List<BedEntity> {
+    val beds = mutableListOf<BedEntity>()
+    for (i in 1..<dataFrame.columnsCount()) {
+      val roomAnswers = dataFrame.getColumn(i)
+      val bedNumber = roomAnswers[1].toString()
+      beds.add(
+        createBed(
+          "${roomAnswers.name}",
+          "${roomAnswers.name} - $bedNumber",
+          rooms.firstOrNull { it.code == roomAnswers.name }
+            ?: throw IllegalArgumentException("Room not found with id ${roomAnswers.name} and code ${roomAnswers.name}."),
+        ),
```

Comment:
can you please add function param names here so it's clear what each value represents

Reply:
Added parameter names.